File: typescript/1_classes-objects/pastry.ts

class Pastry {
    public name: string;              // public (default anyway)
    private price: number;
    private ingredients: string[];

    constructor(name: string, price: number, ingredients: string[]) {
        this.name = name;
        this.price = price;
        this.ingredients = ingredients;
    }

    public displayInfo(): void {
        console.log(`Pastry: ${this.name} | Price: $${this.price}`);
    }

    public getPrice(): number {
        return this.price;
    }
}

class Customer {
    constructor(public customerName: string) {}

    orderPastry(pastry: Pastry): void {
        console.log(`${this.customerName} orders a ${pastry.name}`);
        console.log(`That will be $${pastry.getPrice()}`);
    }
}

// Demo
const croissant = new Pastry("Croissant", 4.50, ["flour", "butter"]);
const alice = new Customer("Alice");
croissant.displayInfo();
alice.orderPastry(croissant);

File: python/1_classes-objects/Pastry.py

class Pastry:
    """Represents a pastry with a name, price, and ingredients."""

    def __init__(self, name: str, price: float, ingredients: list[str]):
        """Initializes the Pastry object."""
        self.name = name          # Public: anyone can access
        # Python uses a single leading underscore convention to indicate
        # that an attribute is private/protected, but it is not strictly enforced.
        self._price = price       
        self._ingredients = ingredients 

    def display_info(self) -> None:
        """Prints the pastry's basic information."""
        # Ingredients remain "hidden" (conventionally private) outside
        print(f"Pastry: {self.name} | Price: ${self._price:.2f}")

    @property
    def price(self) -> float:
        """Getter (using property decorator) for the private price field."""
        return self._price


